packages/core/src/render3/i18n/i18n_hydration.ts

```diff
@@ -0,0 +1,252 @@
+    lView: LView, index: number, processTextNode: (node: RNode) => void): Array<number>|null {
+  return _trySerializeI18nBlockImpl(lView, index, processTextNode);
+}
+
+function prepareI18nBlockForHydrationImpl(lView: LView, index: number) {
+  if (!isI18nHydrationSupportEnabled()) {
+    return;
+  }
+
+  const hydrationInfo = lView[HYDRATION];
+  if (!hydrationInfo) {
+    return;
+  }
+
+  const tView = lView[TVIEW];
+  const tI18n = tView.data[index] as TI18n;
+
+  ngDevMode && assertDefined(tI18n, 'Expected i18n data');
+
+  const firstNode = tI18n.ast[0];
+  if (!firstNode) {
+    return;
+  }
+
+  const tNode = tView.data[firstNode.index] as TNode;
+  const rootNode: Node = locateNextRNode(hydrationInfo, tView, lView, tNode) as Node;
+
+  ngDevMode && assertDefined(rootNode, 'expected root node');
+
+  const disconnectedNodes = hydrationInfo.disconnectedNodes ??=
+      new Set(hydrationInfo.data[DISCONNECTED_NODES] ?? []);
+  const i18nMap: Map<number, Node> = hydrationInfo.i18nNodes ??= new Map();
+  const caseQueue: number[] = hydrationInfo.data[I18N_DATA]?.[index - HEADER_OFFSET] ?? [];
+
+  interface HydrationState {
+    currentNode: Node|null;
+  }
+
+  const enum MarkOptions {
+    NONE = 0,
+    CLAIM,
+  }
+
+  function markHydrationRoot(node: I18nNode, state: HydrationState|null, options: MarkOptions) {
+    const noOffsetIndex = node.index - HEADER_OFFSET;
+    const domNode = state?.currentNode;
+
+    if (domNode) {
+      i18nMap.set(noOffsetIndex, domNode);
+      if (options === MarkOptions.CLAIM) {
+        state.currentNode = domNode.nextSibling;
+      }
+
+      disconnectedNodes.delete(noOffsetIndex);
+    } else {
+      disconnectedNodes.add(noOffsetIndex);
+    }
+  }
+
+  function prepareForHydration(nodeOrNodes: I18nNode|I18nNode[], state: HydrationState|null) {
+    if (Array.isArray(nodeOrNodes)) {
+      for (let i = 0; i < nodeOrNodes.length; i++) {
+        prepareForHydration(nodeOrNodes[i], state);
+      }
+    } else {
+      switch (nodeOrNodes.kind) {
+        case I18nNodeKind.TEXT:
+          // Claim a text node for hydration
+          markHydrationRoot(nodeOrNodes, state, MarkOptions.CLAIM);
+          break;
+
+        case I18nNodeKind.ELEMENT:
+          // Recurse into the current element's children...
+          const childState = state ? {currentNode: state.currentNode?.firstChild ?? null} : null;
+          prepareForHydration(nodeOrNodes.children, childState);
+
+          // And claim the parent element itself.
+          markHydrationRoot(nodeOrNodes, state, MarkOptions.CLAIM);
+          break;
+
+        case I18nNodeKind.PLACEHOLDER:
+          const containerSize =
+              getNgContainerSize(hydrationInfo!, nodeOrNodes.index - HEADER_OFFSET);
+
+          switch (nodeOrNodes.type) {
+            case I18nPlaceholderType.ELEMENT:
+              let childState = state;
+              let markOptions = MarkOptions.NONE;
+
+              if (containerSize == null) {
```

Comment:
Do we always have `null` here or we may have `undefined`?
```suggestion
              if (containerSize === null) {
```